### PR TITLE
Allow `Raster`s as arguments -> return types for terrain attributes

### PR DESCRIPTION
**Pull request by erikmannerfelt**
Tackles #178 

Now, this works:
```python
import xdem

dem = xdem.DEM(xdem.examples.get_path("longyearbyen_ref_dem"))

slope = xdem.terrain.slope(dem)  # This is a Raster since a RasterType was given as an argument.

slope.save(...)

```

### CODE REVIEW

File: xdem/terrain.py [hunk 1]
```diff
@@ -463,14 +530,31 @@ def aspect(dem: np.ndarray | np.ma.masked_array, degrees: bool = True) -> np.nda
...
+) -> np.ndarray: ...
+
+def hillshade(
+    dem: np.ndarray | np.ma.masked_array,
+    resolution: float | tuple[float, float] | None = None,
     azimuth: float = 315.0,
     altitude: float = 45.0,
     z_factor: float = 1.0,
-) -> np.ndarray:
+) -> np.ndarray | Raster:
```
Comment (rhugonnet) [inline xdem/terrain.py:557]: how come for hillshade it can be a `Raster` wth a `np.ndarray` as input?

Comment (rhugonnet) [reply xdem/terrain.py:557]: my bad, this is not the `@overload` !

File: xdem/terrain.py [hunk 2]
```diff
@@ -414,12 +454,28 @@ def get_terrain_attribute(
 
     output_attributes = [terrain_attributes[key].reshape(dem.shape) for key in attribute]
 
+    if isinstance(dem, gu.Raster):
```
Comment (rhugonnet) [inline xdem/terrain.py:457]: we're going to write this line quite often I think! but it's not too much, it's fine :smile: !

**Review by rhugonnet — APPROVED** [review]
Perfect :+1:

File: xdem/terrain.py [hunk 3]
```diff
@@ -209,26 +213,56 @@ def get_terrain_attribute(
...
+    degrees: bool,
+    hillshade_altitude: float,
+    hillshade_azimuth: float,
+    hillshade_z_factor: float,
+    fill_method: str,
+    edge_method: str
 ) -> list[np.ndarray]:
     ...
 
+@overload
```
Comment (adehecq) [inline xdem/terrain.py:226]: Those overloads are becoming increasingly complex!

**Comment by adehecq** [comment]
Nice addition! 
But all those overloads don't make the code look simpler :sweat_smile:

**Pull request merged**